interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,1001 @@
+// addRulePermissionToTree adds all the path pattern variants for the given
+// rule to the map for the given permission.
+//
+// If there are conflicting pattern variants from other non-expired rules,
+// all variants which were previously added during this function call are
+// removed from the path map, and an error is returned along with a list of
+// the conflicting variants and the IDs of the conflicting rules.
+//
+// Conflicts with expired rules, however, result in the expired rule being
+// immediately removed, and the new rule can continue to be added.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) addRulePermissionToTree(rule *Rule, permission string) *ruleConflictError {
+	permVariants := rdb.ensurePermissionDBForUserSnapInterfacePermission(rule.User, rule.Snap, rule.Interface, permission)
+
+	newVariantEntries := make(map[string]variantEntry, rule.Constraints.PathPattern.NumVariants())
+	expiredRules := make(map[prompting.IDType]bool)
+	var conflicts []ruleConflict
+
+	addVariant := func(index int, variant patterns.PatternVariant) {
+		newEntry := variantEntry{
+			Variant: variant,
+			RuleID:  rule.ID,
+		}
+		variantStr := variant.String()
+		conflictingVariantEntry, exists := permVariants.VariantEntries[variantStr]
+		switch {
+		case !exists:
+			newVariantEntries[variantStr] = newEntry
+		case rdb.isRuleExpired(conflictingVariantEntry.RuleID, rule.Timestamp):
+			expiredRules[conflictingVariantEntry.RuleID] = true
+			newVariantEntries[variantStr] = newEntry
+		default:
+			// Exists and is not expired, so there's a conflict
+			conflicts = append(conflicts, ruleConflict{
+				Variant:       variantStr,
+				ConflictingID: conflictingVariantEntry.RuleID,
+			})
+		}
+	}
+	rule.Constraints.PathPattern.RenderAllVariants(addVariant)
+
+	if len(conflicts) > 0 {
+		err := &ruleConflictError{
+			conflicts:  conflicts,
+			permission: permission,
+		}
+		return err
+	}
+
+	for ruleID := range expiredRules {
+		removedRule, err := rdb.removeRuleWithID(ruleID)
+		// Error shouldn't occur. If it does, the rule was already removed.
+		if err == nil {
+			rdb.removeRuleFromTree(removedRule)
+			rdb.notifyRule(removedRule.User, removedRule.ID,
+				map[string]string{"removed": "expired"})
+		}
+	}
+
+	for variantStr, entry := range newVariantEntries {
+		permVariants.VariantEntries[variantStr] = entry
+	}
+
+	return nil
+}
+
+// isRuleExpired returns true if the rule with given ID is expired with respect
+// to the provided timestamp, or if it otherwise no longer exists.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) isRuleExpired(id prompting.IDType, currTime time.Time) bool {
+	rule, err := rdb.ruleWithID(id)
+	if err != nil {
+		return true
+	}
+	expired, err := rule.Expired(currTime)
+	if err != nil {
+		// Issue with expiration, this should not occur
+		logger.Noticef("internal error: while checking whether rule had expired: %v", err)
+	}
```

Comment:
a reason why we should stop having Expired return errors, this is just annoying code